processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -31,25 +36,180 @@
 import java.nio.channels.WritableByteChannel;
 import java.util.Arrays;
 import java.util.Iterator;
+import java.util.List;
 
 /**
  * A generic, flat storage mechanism.  Use static methods fromArray() or fromIterable() to construct.  If input
  * is sorted, supports binary search index lookups.  If input is not sorted, only supports array-like index lookups.
- *
+ * <p>
  * V1 Storage Format:
- *
+ * <p>
  * byte 1: version (0x1)
  * byte 2 == 0x1 =&gt; allowReverseLookup
  * bytes 3-6 =&gt; numBytesUsed
  * bytes 7-10 =&gt; numElements
  * bytes 10-((numElements * 4) + 10): integers representing *end* offsets of byte serialized values
  * bytes ((numElements * 4) + 10)-(numBytesUsed + 2): 4-byte integer representing length of value, followed by bytes for value
+ * <p>
+ * V2 Storage Format
+ * Meta, header and value files are separate.
+ * Meta File:
+ * byte 1: version (0x2)
+ * byte 2 == 0x1 =&gt; allowReverseLookup
+ * bytes 3-6: numberOfElementsPerValueFile expressed as power of 2. That means all the value files contains same number of items
+ * except last value file and may have fewer elements.
+ * bytes 7-10 =&gt; numBytesUsed
+ * bytes 11-14 =&gt; numElements
+ * bytes 15-18 =&gt; columnNameLength.
+ * bytes 19-columnNameLength =&gt; columnName
+ * <p>
+ * Header file name is identified as:  String.format("%s_header", columnName)
+ * value files are identified as: String.format("%s_value_%d", columnName, fileNumber)
+ * number of value files == numElements/numberOfElementsPerValueFile
  */
 public class GenericIndexed<T> implements Indexed<T>
 {
-  private static final byte version = 0x1;
+  public static final ObjectStrategy<String> STRING_STRATEGY = new CacheableObjectStrategy<String>()
+  {
+    private final Ordering ordering = Ordering.natural().nullsFirst();
```

Comment:
Could be static. Also there is `GroupByQuery.NATURAL_NULLS_FIRST`, maybe makes sense to extract a common utility.